{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fetch the datasets table"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    },
    "ExecuteTime": {
     "end_time": "2025-02-15T01:37:22.349102Z",
     "start_time": "2025-02-15T01:37:13.195404Z"
    }
   },
   "source": [
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma()\n",
    "census_datasets = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()\n",
    "\n",
    "# for convenience, indexing on the soma_joinid which links this to other census data.\n",
    "# Indexing by soma_joinid breaks slicing dataset by soma_joinid\n",
    "census_datasets = census_datasets.set_index(\"soma_joinid\")\n",
    "\n",
    "# Filter to only include the \"Human Brian Cell Atlas v1.0\" by the collection id == 283d65eb-dd53-496d-adb7-7570c7caa443\n",
    "# census_datasets[census_datasets.collection_id == \"283d65eb-dd53-496d-adb7-7570c7caa443\"]\n",
    "\n",
    "# Filter to only include the Microglia Supercluster by the dataset id == 700aed19-c16e-4ba8-9191-07da098a8626\n",
    "# census_datasets[census_datasets.dataset_id == \"700aed19-c16e-4ba8-9191-07da098a8626\"]\n",
    "\n",
    "census_datasets"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The \"stable\" release is currently 2025-01-30. Specify 'census_version=\"2025-01-30\"' in future calls to open_soma() to ensure data consistency.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "                                                      citation  \\\n",
       "soma_joinid                                                      \n",
       "0            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "1            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "2            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "3            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "4            Publication: https://doi.org/10.1038/s41591-02...   \n",
       "...                                                        ...   \n",
       "1568         Publication: https://doi.org/10.1038/s41586-02...   \n",
       "1569         Publication: https://doi.org/10.1126/science.a...   \n",
       "1570         Publication: https://doi.org/10.1038/s41586-02...   \n",
       "1571         Publication: https://doi.org/10.1038/s41593-02...   \n",
       "1572         Publication: https://doi.org/10.1038/s41593-02...   \n",
       "\n",
       "                                    collection_id  \\\n",
       "soma_joinid                                         \n",
       "0            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "1            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "2            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "3            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "4            a96133de-e951-4e2d-ace6-59db8b3bfb1d   \n",
       "...                                           ...   \n",
       "1568         45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
       "1569         e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "1570         45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
       "1571         1ca90a2d-2943-483d-b678-b809bf464c30   \n",
       "1572         1ca90a2d-2943-483d-b678-b809bf464c30   \n",
       "\n",
       "                                               collection_name  \\\n",
       "soma_joinid                                                      \n",
       "0            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "1            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "2            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "3            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "4            HTAN/HTAPP Broad - Spatio-molecular dissection...   \n",
       "...                                                        ...   \n",
       "1568         A single-cell transcriptional timelapse of mou...   \n",
       "1569                                            Tabula Sapiens   \n",
       "1570         A single-cell transcriptional timelapse of mou...   \n",
       "1571         SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
       "1572         SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
       "\n",
       "                         collection_doi  \\\n",
       "soma_joinid                               \n",
       "0            10.1016/j.isci.2022.104097   \n",
       "1            10.1016/j.isci.2022.104097   \n",
       "2            10.1016/j.isci.2022.104097   \n",
       "3            10.1016/j.isci.2022.104097   \n",
       "4            10.1038/s41591-024-03215-z   \n",
       "...                                 ...   \n",
       "1568         10.1038/s41586-024-07069-w   \n",
       "1569            10.1126/science.abl4896   \n",
       "1570         10.1038/s41586-024-07069-w   \n",
       "1571         10.1038/s41593-024-01774-5   \n",
       "1572         10.1038/s41593-024-01774-5   \n",
       "\n",
       "                                          collection_doi_label  \\\n",
       "soma_joinid                                                      \n",
       "0                              Marshall et al. (2022) iScience   \n",
       "1                              Marshall et al. (2022) iScience   \n",
       "2                              Marshall et al. (2022) iScience   \n",
       "3                              Marshall et al. (2022) iScience   \n",
       "4                             Klughammer et al. (2024) Nat Med   \n",
       "...                                                        ...   \n",
       "1568                                  Qiu et al. (2024) Nature   \n",
       "1569         The Tabula Sapiens Consortium* et al. (2022) S...   \n",
       "1570                                  Qiu et al. (2024) Nature   \n",
       "1571                        Gabitto et al. (2024) Nat Neurosci   \n",
       "1572                        Gabitto et al. (2024) Nat Neurosci   \n",
       "\n",
       "                                       dataset_id  \\\n",
       "soma_joinid                                         \n",
       "0            4eb29386-de81-452f-b3c0-e00844e8c7fd   \n",
       "1            78d59e4a-82eb-4a61-a1dc-da974d7ea54b   \n",
       "2            add5eb84-5fc9-4f01-982e-a346dd42ee82   \n",
       "3            b020294c-ab82-4547-b5a7-63d8ffa575ed   \n",
       "4            d7476ae2-e320-4703-8304-da5c42627e71   \n",
       "...                                           ...   \n",
       "1568         dcfa2614-7ca7-4d82-814c-350626eccb26   \n",
       "1569         53d208b0-2cfd-4366-9866-c3c6114081bc   \n",
       "1570         dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3   \n",
       "1571         c2876b1b-06d8-4d96-a56b-5304f815b99a   \n",
       "1572         6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3   \n",
       "\n",
       "                               dataset_version_id  \\\n",
       "soma_joinid                                         \n",
       "0            f76861bb-becb-4eb7-82fc-782dc96ccc7f   \n",
       "1            7d7ec1b6-6e3f-4aaa-9442-4b22f3424396   \n",
       "2            de54aed8-4f73-48f6-9229-418a840e2d82   \n",
       "3            abe4fce1-0859-4a56-ad1e-734d79f0e6c8   \n",
       "4            863fc5e4-bd4a-4681-9c3d-0ee7ef54e327   \n",
       "...                                           ...   \n",
       "1568         3002a659-a1a9-4406-9976-99e658e1fbb5   \n",
       "1569         10df7690-6d10-4029-a47e-0f071bb2df83   \n",
       "1570         3817734b-0f82-433b-8c38-55b214200fff   \n",
       "1571         c32964d2-3339-441f-8e56-7177234c7876   \n",
       "1572         d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b   \n",
       "\n",
       "                                                 dataset_title  \\\n",
       "soma_joinid                                                      \n",
       "0             Spatial transcriptomics in mouse: Puck_191112_05   \n",
       "1             Spatial transcriptomics in mouse: Puck_191112_08   \n",
       "2             Spatial transcriptomics in mouse: Puck_191109_20   \n",
       "3             Spatial transcriptomics in mouse: Puck_191112_13   \n",
       "4                                 HTAPP-330-SMP-1082 scRNA-seq   \n",
       "...                                                        ...   \n",
       "1568                              Major cell cluster: Mesoderm   \n",
       "1569                                Tabula Sapiens - All Cells   \n",
       "1570                            Whole dataset: Raw counts only   \n",
       "1571         Whole Taxonomy - MTG: Seattle Alzheimer's Dise...   \n",
       "1572         Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...   \n",
       "\n",
       "                                     dataset_h5ad_path  \\\n",
       "soma_joinid                                              \n",
       "0            4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad   \n",
       "1            78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad   \n",
       "2            add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad   \n",
       "3            b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad   \n",
       "4            d7476ae2-e320-4703-8304-da5c42627e71.h5ad   \n",
       "...                                                ...   \n",
       "1568         dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad   \n",
       "1569         53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad   \n",
       "1570         dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad   \n",
       "1571         c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad   \n",
       "1572         6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad   \n",
       "\n",
       "             dataset_total_cell_count  \n",
       "soma_joinid                            \n",
       "0                               10888  \n",
       "1                               10250  \n",
       "2                               12906  \n",
       "3                               15161  \n",
       "4                                 565  \n",
       "...                               ...  \n",
       "1568                          3267338  \n",
       "1569                          1136218  \n",
       "1570                         11441407  \n",
       "1571                          1226855  \n",
       "1572                          1309414  \n",
       "\n",
       "[1573 rows x 10 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>citation</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>collection_doi_label</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_version_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>soma_joinid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd</td>\n",
       "      <td>f76861bb-becb-4eb7-82fc-782dc96ccc7f</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_05</td>\n",
       "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad</td>\n",
       "      <td>10888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b</td>\n",
       "      <td>7d7ec1b6-6e3f-4aaa-9442-4b22f3424396</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_08</td>\n",
       "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad</td>\n",
       "      <td>10250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82</td>\n",
       "      <td>de54aed8-4f73-48f6-9229-418a840e2d82</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_20</td>\n",
       "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad</td>\n",
       "      <td>12906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed</td>\n",
       "      <td>abe4fce1-0859-4a56-ad1e-734d79f0e6c8</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_13</td>\n",
       "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad</td>\n",
       "      <td>15161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41591-02...</td>\n",
       "      <td>a96133de-e951-4e2d-ace6-59db8b3bfb1d</td>\n",
       "      <td>HTAN/HTAPP Broad - Spatio-molecular dissection...</td>\n",
       "      <td>10.1038/s41591-024-03215-z</td>\n",
       "      <td>Klughammer et al. (2024) Nat Med</td>\n",
       "      <td>d7476ae2-e320-4703-8304-da5c42627e71</td>\n",
       "      <td>863fc5e4-bd4a-4681-9c3d-0ee7ef54e327</td>\n",
       "      <td>HTAPP-330-SMP-1082 scRNA-seq</td>\n",
       "      <td>d7476ae2-e320-4703-8304-da5c42627e71.h5ad</td>\n",
       "      <td>565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1568</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
       "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "      <td>10.1038/s41586-024-07069-w</td>\n",
       "      <td>Qiu et al. (2024) Nature</td>\n",
       "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26</td>\n",
       "      <td>3002a659-a1a9-4406-9976-99e658e1fbb5</td>\n",
       "      <td>Major cell cluster: Mesoderm</td>\n",
       "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad</td>\n",
       "      <td>3267338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569</th>\n",
       "      <td>Publication: https://doi.org/10.1126/science.a...</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>The Tabula Sapiens Consortium* et al. (2022) S...</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>\n",
       "      <td>10df7690-6d10-4029-a47e-0f071bb2df83</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad</td>\n",
       "      <td>1136218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1570</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
       "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "      <td>10.1038/s41586-024-07069-w</td>\n",
       "      <td>Qiu et al. (2024) Nature</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
       "      <td>3817734b-0f82-433b-8c38-55b214200fff</td>\n",
       "      <td>Whole dataset: Raw counts only</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad</td>\n",
       "      <td>11441407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1571</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
       "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
       "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
       "      <td>10.1038/s41593-024-01774-5</td>\n",
       "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
       "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a</td>\n",
       "      <td>c32964d2-3339-441f-8e56-7177234c7876</td>\n",
       "      <td>Whole Taxonomy - MTG: Seattle Alzheimer's Dise...</td>\n",
       "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad</td>\n",
       "      <td>1226855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1572</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
       "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
       "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
       "      <td>10.1038/s41593-024-01774-5</td>\n",
       "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
       "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3</td>\n",
       "      <td>d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b</td>\n",
       "      <td>Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...</td>\n",
       "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad</td>\n",
       "      <td>1309414</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573 rows × 10 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 8
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify all the genes measured in the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>feature_length</th>\n",
       "      <th>nnz</th>\n",
       "      <th>n_measured_obs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1059</td>\n",
       "      <td>7958785</td>\n",
       "      <td>92252850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1244</td>\n",
       "      <td>18685092</td>\n",
       "      <td>105784525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000187642</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2765</td>\n",
       "      <td>664016</td>\n",
       "      <td>95688802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000272512</td>\n",
       "      <td>ENSG00000272512.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2086</td>\n",
       "      <td>927207</td>\n",
       "      <td>91649589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000188290</td>\n",
       "      <td>HES4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>961</td>\n",
       "      <td>19206715</td>\n",
       "      <td>105542421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60383</th>\n",
       "      <td>60383</td>\n",
       "      <td>ENSG00000285470</td>\n",
       "      <td>TUSC2P1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1237</td>\n",
       "      <td>2150</td>\n",
       "      <td>23175703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60384</th>\n",
       "      <td>60384</td>\n",
       "      <td>ENSG00000286120</td>\n",
       "      <td>ENSG00000286120.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>457</td>\n",
       "      <td>6</td>\n",
       "      <td>19614003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60385</th>\n",
       "      <td>60385</td>\n",
       "      <td>ENSG00000286173</td>\n",
       "      <td>ENSG00000286173.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>457</td>\n",
       "      <td>3</td>\n",
       "      <td>19614003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60386</th>\n",
       "      <td>60386</td>\n",
       "      <td>ENSG00000286744</td>\n",
       "      <td>ENSG00000286744.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>102</td>\n",
       "      <td>74</td>\n",
       "      <td>19229873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60426</th>\n",
       "      <td>60426</td>\n",
       "      <td>ENSG00000289084</td>\n",
       "      <td>ENSG00000289084.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>637</td>\n",
       "      <td>4845634</td>\n",
       "      <td>20936510</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59236 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid       feature_id       feature_name  \\\n",
       "0                0  ENSG00000237491          LINC01409   \n",
       "1                1  ENSG00000188976              NOC2L   \n",
       "2                2  ENSG00000187642              PERM1   \n",
       "3                3  ENSG00000272512  ENSG00000272512.1   \n",
       "4                4  ENSG00000188290               HES4   \n",
       "...            ...              ...                ...   \n",
       "60383        60383  ENSG00000285470            TUSC2P1   \n",
       "60384        60384  ENSG00000286120  ENSG00000286120.1   \n",
       "60385        60385  ENSG00000286173  ENSG00000286173.1   \n",
       "60386        60386  ENSG00000286744  ENSG00000286744.1   \n",
       "60426        60426  ENSG00000289084  ENSG00000289084.1   \n",
       "\n",
       "                             feature_type  feature_length       nnz  \\\n",
       "0                                  lncRNA            1059   7958785   \n",
       "1                          protein_coding            1244  18685092   \n",
       "2                          protein_coding            2765    664016   \n",
       "3                                  lncRNA            2086    927207   \n",
       "4                          protein_coding             961  19206715   \n",
       "...                                   ...             ...       ...   \n",
       "60383                processed_pseudogene            1237      2150   \n",
       "60384  transcribed_unprocessed_pseudogene             457         6   \n",
       "60385  transcribed_unprocessed_pseudogene             457         3   \n",
       "60386              unprocessed_pseudogene             102        74   \n",
       "60426                              lncRNA             637   4845634   \n",
       "\n",
       "       n_measured_obs  \n",
       "0            92252850  \n",
       "1           105784525  \n",
       "2            95688802  \n",
       "3            91649589  \n",
       "4           105542421  \n",
       "...               ...  \n",
       "60383        23175703  \n",
       "60384        19614003  \n",
       "60385        19614003  \n",
       "60386        19229873  \n",
       "60426        20936510  \n",
       "\n",
       "[59236 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the experiment containing human data, and the measurement therein with RNA\n",
    "human = census[\"census_data\"][\"homo_sapiens\"]\n",
    "human_rna = human.ms[\"RNA\"]\n",
    "\n",
    "# The census-wide datasets\n",
    "datasets_df = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()\n",
    "\n",
    "presence_matrix = cellxgene_census.get_presence_matrix(census, organism=\"Homo sapiens\", measurement_name=\"RNA\")\n",
    "\n",
    "var_df = human_rna.var.read().concat().to_pandas()\n",
    "\n",
    "# Slice the dataset(s) of interest, and get the joinid(s)\n",
    "dataset_joinids = datasets_df.loc[datasets_df.dataset_id == \"700aed19-c16e-4ba8-9191-07da098a8626\"].soma_joinid\n",
    "\n",
    "# Slice the presence matrix by the first dimension, i.e., by dataset\n",
    "var_joinids = presence_matrix[dataset_joinids, :].tocoo().col\n",
    "\n",
    "# From the feature (var) dataframe, slice out features which have a joinid in the list.\n",
    "var_df.loc[var_df.soma_joinid.isin(var_joinids)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize the gene expression?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  },
  "polyglot_notebook": {
   "kernelInfo": {
    "defaultKernelName": "csharp",
    "items": [
     {
      "aliases": [],
      "name": "csharp"
     }
    ]
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
