{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Querying data using the gget cellxgene module\n",
    "\n",
    "gget is a free, open-source command-line tool and Python package that enables efficient querying of genomic databases. gget consists of a collection of separate but interoperable modules, each designed to facilitate one type of database querying in a single line of code."
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-15T00:27:37.647414Z",
     "start_time": "2025-02-15T00:27:34.651588Z"
    }
   },
   "source": [
    "import gget\n",
    "gget.setup(\"cellxgene\")\n",
    "\n",
    "# Display all options of the cellxgene gget module\n",
    "help(gget.cellxgene)"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "19:27:35 - INFO - Installing cellxgene-census package (requires pip).\n",
      "19:27:37 - INFO - cellxgene_census installed succesfully.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function cellxgene in module gget.gget_cellxgene:\n",
      "\n",
      "cellxgene(species='homo_sapiens', gene=None, ensembl=False, column_names=['dataset_id', 'assay', 'suspension_type', 'sex', 'tissue_general', 'tissue', 'cell_type'], meta_only=False, tissue=None, cell_type=None, development_stage=None, disease=None, sex=None, is_primary_data=True, dataset_id=None, tissue_general_ontology_term_id=None, tissue_general=None, assay_ontology_term_id=None, assay=None, cell_type_ontology_term_id=None, development_stage_ontology_term_id=None, disease_ontology_term_id=None, donor_id=None, self_reported_ethnicity_ontology_term_id=None, self_reported_ethnicity=None, sex_ontology_term_id=None, suspension_type=None, tissue_ontology_term_id=None, census_version='stable', verbose=True, out=None)\n",
      "    Query data from CZ CELLxGENE Discover (https://cellxgene.cziscience.com/) using the\n",
      "    CZ CELLxGENE Discover Census (https://github.com/chanzuckerberg/cellxgene-census).\n",
      "\n",
      "    NOTE: Querying large datasets requires a large amount of RAM. Use the cell metadata attributes\n",
      "    to define the (sub)dataset of interest.\n",
      "    The CZ CELLxGENE Discover Census recommends >16 GB of memory and a >5 Mbps internet connection.\n",
      "\n",
      "    General args:\n",
      "        - species        Choice of 'homo_sapiens' or 'mus_musculus'. Default: 'homo_sapiens'.\n",
      "        - gene           Str or list of gene name(s) or Ensembl ID(s), e.g. ['ACE2', 'SLC5A1'] or ['ENSG00000130234', 'ENSG00000100170']. Default: None.\n",
      "                         NOTE: Set ensembl=True when providing Ensembl ID(s) instead of gene name(s).\n",
      "                         See https://cellxgene.cziscience.com/gene-expression for examples of available genes.\n",
      "        - ensembl        True/False (default: False). Set to True when genes are provided as Ensembl IDs.\n",
      "        - column_names   List of metadata columns to return (stored in AnnData.obs when meta_only=False).\n",
      "                         Default: [\"dataset_id\", \"assay\", \"suspension_type\", \"sex\", \"tissue_general\", \"tissue\", \"cell_type\"]\n",
      "                         For more options see: https://api.cellxgene.cziscience.com/curation/ui/#/ -> Schemas -> dataset\n",
      "        - meta_only      True/False (default: False). If True, returns only metadata dataframe (corresponds to AnnData.obs).\n",
      "        - census_version Str defining version of Census, e.g. \"2023-05-15\" or \"latest\" or \"stable\". Default: \"stable\".\n",
      "        - verbose        True/False whether to print progress information. Default True.\n",
      "        - out            If provided, saves the generated AnnData h5ad (or csv when meta_only=True) file with the specified path. Default: None.\n",
      "\n",
      "    Cell metadata attributes:\n",
      "        - tissue                          Str or list of tissue(s), e.g. ['lung', 'blood']. Default: None.\n",
      "                                          See https://cellxgene.cziscience.com/gene-expression for examples of available tissues.\n",
      "        - cell_type                       Str or list of celltype(s), e.g. ['mucus secreting cell', 'neuroendocrine cell']. Default: None.\n",
      "                                          See https://cellxgene.cziscience.com/gene-expression and select a tissue to see examples of available celltypes.\n",
      "        - development_stage               Str or list of development stage(s). Default: None.\n",
      "        - disease                         Str or list of disease(s). Default: None.\n",
      "        - sex                             Str or list of sex(es), e.g. 'female'. Default: None.\n",
      "        - is_primary_data                 True/False (default: True). If True, returns only the canonical instance of the cellular observation.\n",
      "                                          This is commonly set to False for meta-analyses reusing data or for secondary views of data.\n",
      "        - dataset_id                      Str or list of CELLxGENE dataset ID(s). Default: None.\n",
      "        - tissue_general_ontology_term_id Str or list of high-level tissue UBERON ID(s). Default: None.\n",
      "                                          Also see: https://github.com/chanzuckerberg/single-cell-data-portal/blob/9b94ccb0a2e0a8f6182b213aa4852c491f6f6aff/backend/wmg/data/tissue_mapper.py\n",
      "        - tissue_general                  Str or list of high-level tissue label(s). Default: None.\n",
      "                                          Also see: https://github.com/chanzuckerberg/single-cell-data-portal/blob/9b94ccb0a2e0a8f6182b213aa4852c491f6f6aff/backend/wmg/data/tissue_mapper.py\n",
      "        - tissue_ontology_term_id         Str or list of tissue ontology term ID(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - assay_ontology_term_id          Str or list of assay ontology term ID(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - assay                           Str or list of assay(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - cell_type_ontology_term_id      Str or list of celltype ontology term ID(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - development_stage_ontology_term_id        Str or list of development stage ontology term ID(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - disease_ontology_term_id        Str or list of disease ontology term ID(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - donor_id                        Str or list of donor ID(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - self_reported_ethnicity_ontology_term_id  Str or list of self reported ethnicity ontology ID(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - self_reported_ethnicity         Str or list of self reported ethnicity as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - sex_ontology_term_id            Str or list of sex ontology ID(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "        - suspension_type                 Str or list of suspension type(s) as defined in the CELLxGENE dataset schema. Default: None.\n",
      "\n",
      "    Returns AnnData object (when meta_only=False) or dataframe (when meta_only=True).\n",
      "\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Fetch an AnnData object by selecting gene(s), tissue(s) and cell type(s)"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-15T01:52:17.743676Z",
     "start_time": "2025-02-15T01:45:09.792654Z"
    }
   },
   "cell_type": "code",
   "source": [
    "adata_homo = gget.cellxgene(\n",
    "    meta_only=False,\n",
    "    ensembl=False,  # Setting 'ensembl=True' here since the gene is passed as an Ensembl ID\n",
    "    # collection_id=\"\",\n",
    "    # dataset_id = \"700aed19-c16e-4ba8-9191-07da098a8626\",\n",
    "    # gene=[\"ENSG00000197405\",\"ENSG00000171860\",\"ENSG00000000971\"],\n",
    "    gene=[\"CFH\", \"C5AR1\", \"C3AR1\"],\n",
    "    assay=[\"10x 3' v3\"],\n",
    "    tissue=\"brain\",\n",
    "    # species=\"mus_musculus\",  # Let's switch up the species\n",
    "    cell_type=\"microglial cell\",\n",
    "    disease=\"normal\",\n",
    ")\n",
    "\n",
    "adata_homo"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:45:09 - INFO - Fetching AnnData object from CZ CELLxGENE Discover. This might take a few minutes...\n",
      "The \"stable\" release is currently 2025-01-30. Specify 'census_version=\"2025-01-30\"' in future calls to open_soma() to ensure data consistency.\n",
      "/Users/aumchampaneri/.pyenv/versions/miniconda3-latest/lib/python3.12/site-packages/gget/gget_cellxgene.py:189: FutureWarning: The argument `column_names` is deprecated and will be removed in a future release. Please use `obs_column_names` and `var_column_names` instead.\n",
      "  adata = cellxgene_census.get_anndata(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 12 × 3\n",
       "    obs: 'dataset_id', 'assay', 'suspension_type', 'sex', 'tissue_general', 'tissue', 'cell_type', 'is_primary_data', 'disease'\n",
       "    var: 'soma_joinid', 'feature_id', 'feature_name', 'feature_type', 'feature_length', 'nnz', 'n_measured_obs'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-15T02:18:57.013733Z",
     "start_time": "2025-02-15T02:07:28.962335Z"
    }
   },
   "cell_type": "code",
   "source": [
    "adata_mus = gget.cellxgene(\n",
    "    meta_only=False,\n",
    "    ensembl=False,  # Setting 'ensembl=True' here since the gene is passed as an Ensembl ID\n",
    "    # collection_id=\"\",\n",
    "    # dataset_id = \"700aed19-c16e-4ba8-9191-07da098a8626\",\n",
    "    # gene=[\"ENSG00000197405\",\"ENSG00000171860\",\"ENSG00000000971\"],\n",
    "    # gene=[\"CFH\", \"C5AR1\", \"C3AR1\"],\n",
    "    # assay=[\"10x 3' v3\"],\n",
    "    tissue=\"brain\",\n",
    "    species=\"mus_musculus\",  # Let's switch up the species\n",
    "    cell_type=\"microglial cell\",\n",
    "    disease=\"normal\",\n",
    ")\n",
    "\n",
    "adata_mus"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:07:28 - INFO - Fetching AnnData object from CZ CELLxGENE Discover. This might take a few minutes...\n",
      "The \"stable\" release is currently 2025-01-30. Specify 'census_version=\"2025-01-30\"' in future calls to open_soma() to ensure data consistency.\n",
      "/Users/aumchampaneri/.pyenv/versions/miniconda3-latest/lib/python3.12/site-packages/gget/gget_cellxgene.py:189: FutureWarning: The argument `column_names` is deprecated and will be removed in a future release. Please use `obs_column_names` and `var_column_names` instead.\n",
      "  adata = cellxgene_census.get_anndata(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16609 × 52483\n",
       "    obs: 'dataset_id', 'assay', 'suspension_type', 'sex', 'tissue_general', 'tissue', 'cell_type', 'is_primary_data', 'disease'\n",
       "    var: 'soma_joinid', 'feature_id', 'feature_name', 'feature_type', 'feature_length', 'nnz', 'n_measured_obs'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 30
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-15T02:19:56.421825Z",
     "start_time": "2025-02-15T02:19:56.384002Z"
    }
   },
   "cell_type": "code",
   "source": "adata_mus.obs",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                 dataset_id       assay suspension_type  \\\n",
       "0      3bbb6cf9-72b9-41be-b568-656de6eb18b5   10x 3' v3         nucleus   \n",
       "1      3bbb6cf9-72b9-41be-b568-656de6eb18b5   10x 3' v3         nucleus   \n",
       "2      3bbb6cf9-72b9-41be-b568-656de6eb18b5   10x 3' v3         nucleus   \n",
       "3      3bbb6cf9-72b9-41be-b568-656de6eb18b5   10x 3' v3         nucleus   \n",
       "4      3bbb6cf9-72b9-41be-b568-656de6eb18b5   10x 3' v3         nucleus   \n",
       "...                                     ...         ...             ...   \n",
       "16604  98e5ea9f-16d6-47ec-a529-686e76515e39  Smart-seq2            cell   \n",
       "16605  98e5ea9f-16d6-47ec-a529-686e76515e39  Smart-seq2            cell   \n",
       "16606  98e5ea9f-16d6-47ec-a529-686e76515e39  Smart-seq2            cell   \n",
       "16607  98e5ea9f-16d6-47ec-a529-686e76515e39  Smart-seq2            cell   \n",
       "16608  98e5ea9f-16d6-47ec-a529-686e76515e39  Smart-seq2            cell   \n",
       "\n",
       "          sex tissue_general tissue        cell_type  is_primary_data disease  \n",
       "0      female          brain  brain  microglial cell             True  normal  \n",
       "1      female          brain  brain  microglial cell             True  normal  \n",
       "2      female          brain  brain  microglial cell             True  normal  \n",
       "3      female          brain  brain  microglial cell             True  normal  \n",
       "4      female          brain  brain  microglial cell             True  normal  \n",
       "...       ...            ...    ...              ...              ...     ...  \n",
       "16604    male          brain  brain  microglial cell             True  normal  \n",
       "16605    male          brain  brain  microglial cell             True  normal  \n",
       "16606    male          brain  brain  microglial cell             True  normal  \n",
       "16607    male          brain  brain  microglial cell             True  normal  \n",
       "16608    male          brain  brain  microglial cell             True  normal  \n",
       "\n",
       "[16609 rows x 9 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>assay</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue_general</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3bbb6cf9-72b9-41be-b568-656de6eb18b5</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>female</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3bbb6cf9-72b9-41be-b568-656de6eb18b5</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>female</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3bbb6cf9-72b9-41be-b568-656de6eb18b5</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>female</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3bbb6cf9-72b9-41be-b568-656de6eb18b5</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>female</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3bbb6cf9-72b9-41be-b568-656de6eb18b5</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>female</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16604</th>\n",
       "      <td>98e5ea9f-16d6-47ec-a529-686e76515e39</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>cell</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16605</th>\n",
       "      <td>98e5ea9f-16d6-47ec-a529-686e76515e39</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>cell</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16606</th>\n",
       "      <td>98e5ea9f-16d6-47ec-a529-686e76515e39</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>cell</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16607</th>\n",
       "      <td>98e5ea9f-16d6-47ec-a529-686e76515e39</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>cell</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16608</th>\n",
       "      <td>98e5ea9f-16d6-47ec-a529-686e76515e39</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>cell</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>brain</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16609 rows × 9 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 31
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-15T02:19:58.225083Z",
     "start_time": "2025-02-15T02:19:58.218310Z"
    }
   },
   "cell_type": "code",
   "source": "adata_mus.var",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "       soma_joinid          feature_id feature_name    feature_type  \\\n",
       "0                0  ENSMUSG00000021124        Vti1b  protein_coding   \n",
       "1                1  ENSMUSG00000039377          Hlx  protein_coding   \n",
       "2                2  ENSMUSG00000085604      Dhx58os          lncRNA   \n",
       "3                3  ENSMUSG00000085125      Gm16070          lncRNA   \n",
       "4                4  ENSMUSG00000029439       Sfswap  protein_coding   \n",
       "...            ...                 ...          ...             ...   \n",
       "52478        52478  ENSMUSG00002076971     Snord100          snoRNA   \n",
       "52479        52479  ENSMUSG00002076818     Snord14e          snoRNA   \n",
       "52480        52480  ENSMUSG00002076766      Snord33          snoRNA   \n",
       "52481        52481  ENSMUSG00000118645      Gm55062          lncRNA   \n",
       "52482        52482  ENSMUSG00000118652      Gm54771          lncRNA   \n",
       "\n",
       "       feature_length       nnz  n_measured_obs  \n",
       "0                 841   7020258        42776863  \n",
       "1                1495    481426        42641734  \n",
       "2                1594    117050        40553677  \n",
       "3                2146   2280673        41179548  \n",
       "4                 692  17553003        42776863  \n",
       "...               ...       ...             ...  \n",
       "52478              74      3653           73347  \n",
       "52479              90      3579           73347  \n",
       "52480              84       612           73347  \n",
       "52481             480         0          301796  \n",
       "52482             358         1          301796  \n",
       "\n",
       "[52483 rows x 7 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>feature_length</th>\n",
       "      <th>nnz</th>\n",
       "      <th>n_measured_obs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000021124</td>\n",
       "      <td>Vti1b</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>841</td>\n",
       "      <td>7020258</td>\n",
       "      <td>42776863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000039377</td>\n",
       "      <td>Hlx</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1495</td>\n",
       "      <td>481426</td>\n",
       "      <td>42641734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000085604</td>\n",
       "      <td>Dhx58os</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1594</td>\n",
       "      <td>117050</td>\n",
       "      <td>40553677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000085125</td>\n",
       "      <td>Gm16070</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2146</td>\n",
       "      <td>2280673</td>\n",
       "      <td>41179548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000029439</td>\n",
       "      <td>Sfswap</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>692</td>\n",
       "      <td>17553003</td>\n",
       "      <td>42776863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52478</th>\n",
       "      <td>52478</td>\n",
       "      <td>ENSMUSG00002076971</td>\n",
       "      <td>Snord100</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>74</td>\n",
       "      <td>3653</td>\n",
       "      <td>73347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52479</th>\n",
       "      <td>52479</td>\n",
       "      <td>ENSMUSG00002076818</td>\n",
       "      <td>Snord14e</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>90</td>\n",
       "      <td>3579</td>\n",
       "      <td>73347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52480</th>\n",
       "      <td>52480</td>\n",
       "      <td>ENSMUSG00002076766</td>\n",
       "      <td>Snord33</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>84</td>\n",
       "      <td>612</td>\n",
       "      <td>73347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52481</th>\n",
       "      <td>52481</td>\n",
       "      <td>ENSMUSG00000118645</td>\n",
       "      <td>Gm55062</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>480</td>\n",
       "      <td>0</td>\n",
       "      <td>301796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52482</th>\n",
       "      <td>52482</td>\n",
       "      <td>ENSMUSG00000118652</td>\n",
       "      <td>Gm54771</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>358</td>\n",
       "      <td>1</td>\n",
       "      <td>301796</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>52483 rows × 7 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 32
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
